shell/platform/linux/fl_application.cc

```diff
@@ -0,0 +1,189 @@
+
+#include "flutter/shell/platform/linux/public/flutter_linux/fl_application.h"
+#include "flutter/shell/platform/linux/public/flutter_linux/fl_dart_project.h"
+#include "flutter/shell/platform/linux/public/flutter_linux/fl_view.h"
+
+#include <gmodule.h>
+#include <gtk/gtk.h>
+
+#ifdef GDK_WINDOWING_X11
+#include <gdk/gdkx.h>
+#endif
+
+struct FlApplicationPrivate {
+  char** dart_entrypoint_arguments;
+  gboolean activated;
+};
+#define FL_APPLICATION_PRIVATE(val) static_cast<FlApplicationPrivate*>(val)
+
+enum {
+    VIEW_CONSTRUCTED,
+    LAST_SIGNAL
+};
+
+static guint signals[LAST_SIGNAL] = {0};
+
+G_DEFINE_TYPE_WITH_CODE(FlApplication,
+                        fl_application,
+                        GTK_TYPE_APPLICATION,
+                        G_ADD_PRIVATE(FlApplication))
+
+// Implements GApplication::activate.
+static void fl_application_activate(GApplication* application) {
+  FlApplication* self = FL_APPLICATION(application);
+  FlApplicationPrivate* priv =
+      FL_APPLICATION_PRIVATE(fl_application_get_instance_private(self));
+  
+  GtkApplicationWindow* app_window =
+      GTK_APPLICATION_WINDOW(gtk_application_window_new(GTK_APPLICATION(self)));
+  GtkWindow* window = GTK_WINDOW(app_window);
+
+  // Ensure we only run this function once. We currently don't do anything fancy
+  // like signaling the application that it should show itself again. 
+  // This only stops the same instance of GApplication creating multiple
+  // windows/views.
+  if (priv->activated) {
+    return;
+  }
+  priv->activated = TRUE;
+
+  // Use a header bar when running in GNOME as this is the common style used
+  // by applications and is the setup most users will be using (e.g. Ubuntu
+  // desktop).
+  // If running on X and not using GNOME then just use a traditional title bar
+  // in case the window manager does more exotic layout, e.g. tiling.
+  // If running on Wayland assume the header bar will work (may need changing
+  // if future cases occur).
+  gboolean use_header_bar = TRUE;
+#ifdef GDK_WINDOWING_X11
+  GdkScreen* screen = gtk_window_get_screen(window);
+  if (GDK_IS_X11_SCREEN(screen)) {
+    const gchar* wm_name = gdk_x11_screen_get_window_manager_name(screen);
+    if (g_strcmp0(wm_name, "GNOME Shell") != 0) {
+      use_header_bar = FALSE;
+    }
+  }
+#endif
+  if (use_header_bar) {
+    GtkHeaderBar* header_bar = GTK_HEADER_BAR(gtk_header_bar_new());
+    gtk_header_bar_set_show_close_button(header_bar, true);
+    gtk_widget_show(GTK_WIDGET(header_bar));
+
+    gtk_window_set_titlebar(window, GTK_WIDGET(header_bar));
+  }
+
+  g_autoptr(FlDartProject) project = fl_dart_project_new();
+  fl_dart_project_set_dart_entrypoint_arguments(
+      project, priv->dart_entrypoint_arguments);
+
+  FlView* view = fl_view_new(project);
+  gtk_widget_show(GTK_WIDGET(view));
+  gtk_container_add(GTK_CONTAINER(app_window), GTK_WIDGET(view));
+
+  gtk_widget_grab_focus(GTK_WIDGET(view));
+
+  g_signal_emit (G_OBJECT (self),
```

Comment:
This shouldn't be necessary - a Flutter application can just override the activate method (chaining up to the FlApplication implementation) or connect to the GApplication::activate signal.

Reply:
I've been thinking how this would work for multi-window. I think the user needs the ability to create the window themselves in this signal, as a multi-window app might have special windows that use GTK widgets. Since this is complicated and needs to be well thought out let's drop this from this PR and resolve in follow up PRs.